Add support for the Git dispatcher

michalis1 opened the pull request:
A new 'offline' dispatcher named **GitDefault** has been introduced, which allows collecting raw command outputs provided from a Nautobot Git repository. In this case, the dispatcher framework used is **Git** (instead of Napalm or Netmiko), so the default path for the Git dispatcher will be `nornir_nautobot.plugins.tasks.dispatcher.default.GitDefault`.

michalis1 commented on nornir_nautobot/utils/helpers.py:
```diff
@@ -84,3 +85,54 @@ def get_error_message(error_code: str, **kwargs: Any) -> str:
     return f"{error_code}: {error_message}"
+
+
+def command_to_filename(command):
```
Actually, I haven’t implemented any filename sanitization yet. The user defines a device command string (e.g., show version), and we should generate the filename based on that string (e.g. show_version.raw). The pipe symbol (|) is intended to allow users to specify commands like **show run | sec router bgp**. In such cases, the resulting filename should be **show_run__sec_router_bgp.raw**.

michalis1 replied:
Make sense!!

michalis1 replied:
done

jeffkala commented:
@michalis1 myself and @itdependsnetworks have a few concerns/better options, so I'd say if you want to pause on your work here until we can talk internally.

michalis1 commented on nornir_nautobot/plugins/tasks/dispatcher/default.py:
```diff
@@ -671,6 +672,127 @@ def get_commands(cls, task: Task, logger, obj, command_list, **kwargs):
+            logger.info(f"Reading command output from: {command_file_path}")
+            with command_file_path.open("r", encoding="utf-8") as file:
+                command_raw = file.read()
+            print(command_raw)
```
done

michalis1 commented:
@jeffkala I removed the GitDispatcher and updated **NetmikoDefault** to support offline command retrieval.

Regarding the function arguments, I could use `**kwargs` in **get_command**/**get_commands** instead of defining the individual arguments, but I need to keep them defined in **get_config** since it’s currently not using `**kwargs`—or alternatively, I could add `**kwargs` to **get_config**.

michalis1 commented on nornir_nautobot/utils/helpers.py:
```diff
@@ -84,3 +87,71 @@ def get_error_message(error_code: str, **kwargs: Any) -> str:
+        commands_folder_name = "command_outputs"
+
+    if not command_relative_path:
+        command_relative_path = f"{device.platform.name}/{device.name}/{command_file_name}.raw"
```
That’s why it’s a function parameter—so the app developer can define their own logic and pass it in. In Operational Compliance, we implemented this as a constance setting in a Jinja template form, but I couldn’t follow that pattern here since nornir-nautobot isn’t a Nautobot app and doesn’t have access to settings. So my view was that it’s up to the app developer to implement this setting in their plugin.

jeffkala commented on nornir_nautobot/constants.py:
```diff
@@ -194,6 +194,18 @@
         error_message="Discovered `% Permission denied for the role` in the output",
         recommendation="Ensure that the actual command that is ran, is allowed for the user making the connection. As an example, if `show run` is allowed, but `show running-config` is not, would need to address that.",
     ),
+    "E1031": ErrorCode(
+        troubleshooting="Ensure that the output file for the command `{command}` has been created.",
+        description="While using offline command outputs through Git, the output file for the command `{command}` could not be retrieved.",
+        error_message="The command output file for `{command}` could not be retrieved.",
+        recommendation="Ensure that the output file for the command `{command}` has been created.",
+    ),
+    "E1032": ErrorCode(
+        troubleshooting="Ensure that the output file for the command `{command}` has been created.",
+        description="While using offline command outputs through Git, the output file for the command `{command}` was not found.",
+        error_message="The command output file for `{command}` could not be found.",
+        recommendation="Ensure that the output file for the command `{command}` has been created.",
+    ),
```
Mind updating these to be a bit more distinct?  Appears that 1031 is an issue with loading the file? and 1032 is the file does not exist?  If that is the case can we make the troubleshooting better.

jeffkala commented on nornir_nautobot/plugins/tasks/dispatcher/default.py:
```diff
@@ -605,62 +624,180 @@ def merge_config(  # pylint: disable=too-many-positional-arguments
+
+        command_file_path = get_file_contents_from_git(
+            git_repo_obj=git_repo_obj,
+            # device=obj,
+            # filename=filename,
```
Remove??

michalis1 replied:
done

jeffkala reviewed (approved):
Approved, I would still like @itdependsnetworks to review as well.

@michalis1 a few unittest maybe too even if its just for the helpers?

itdependsnetworks commented on nornir_nautobot/plugins/tasks/dispatcher/default.py:
```diff
@@ -513,6 +515,9 @@ def get_config(  # pylint: disable=too-many-positional-arguments
         backup_file: str,
         remove_lines: list,
         substitute_lines: list,
+        git_repo_obj=None,
+        commands_folder_name=None,
+        command_relative_path=None,
```
Just confirming these will come from GC settings?

michalis1 replied:
Yes, exactly. If Golden Config is going to support this, it should provide a new Git data source specifically for command outputs.

itdependsnetworks replied:
Just making sure it is GC `Settings` though, the emphasis based on this is a single entry point and not per device.

michalis1 replied:
**git_repo_obj** is a Nautobot `GitRepository` object containing the raw command outputs.
**command_folder_name** is the top-level folder in the Git repository, such as "command_outputs".
**command_relative_path** is the path to the command output file relative to `command_folder_name`, for example: `cisco_ios/device_1/show_version.raw`. In Operational Compliance, this has been made configurable via a Constance setting in Jinja template format. For example, the setting for **command_relative_path** could be defined as: "{{ obj.platform }}/{{ obj.name }}/{{ cmd }}.raw". 

Based on the above, the full path to the file containing the raw command output is:
```python
f"{git_repo_obj.filesystem_path}/{command_folder_name}/{command_relative_path}"
```

jeffkala replied:
Provided my thoughts on this internally.

itdependsnetworks commented on nornir_nautobot/plugins/tasks/dispatcher/default.py:
```diff
@@ -605,62 +624,176 @@ def merge_config(  # pylint: disable=too-many-positional-arguments
+        Returns:
+            bool or None:
+                - True or False if the key exists in any of the sources and is explicitly set.
+                - None if the key is not found or does not contain a valid boolean value.
```
When would it return a None?

michalis1 replied:
It wouldn’t. That was from a previous implementation I had done, but I changed it based on Jeff’s comment. I’ll update the docstring accordingly.

itdependsnetworks replied:
ping.

michalis1 replied:
I've already updated this one.

michalis1 replied:
Sorry, I had updated the code but missed the docstring!

itdependsnetworks commented on nornir_nautobot/plugins/tasks/dispatcher/default.py:
```diff
@@ -605,62 +624,176 @@ def merge_config(  # pylint: disable=too-many-positional-arguments
-                    logger.error(error_msg, extra={"object": obj})
-                    raise NornirNautobotException(error_msg)
+                if cls._offline_commands(obj):
+                    command, *rest = command
```
This is a syntax I am not familiar with, but I presume this just unpacks the rest of the list like you would in  function params. 

I am not understanding why we need this?

michalis1 replied:
In this case, we should have the same setting in GoldenConfig, OperationalCompliance, or any other app that needs this feature. However, my main concern is that **nornir-nautobot** is not a Nautobot app, so it doesn’t have access to settings. I was thinking it might not be a good idea to store this jinja2 setting in a custom_field, config_context, or Dispatcher class, like we did for tcp_port.

itdependsnetworks replied:
Yea, that was what I was trying to understand. How should we handle that, and should it be:

* Per nornir job configurable
* Per device configurable
* Per Device & Command Configurable

itdependsnetworks replied:
Finally had time to wrap my brain around this. 

- We should require the task caller to set the file location, not auto generate with `git_repo_obj` and `commands_folder_name`, as that is already the standard we have created. 
    - This is fine in `get_command`
- This does place some complication in `get_commands`, as it is not crazy to say 
    - `{{ obj.platform.name }}/{{ cmd }}/{{ obj.name }}.raw`
- Overall, I don't think we are supporting that anyway, so I think it is fine to only require a folder in `get_commands` and accept that the filename must be based on the name of the command, with optional suffix (presumably)

itdependsnetworks replied:
Spoke with Jeff. We should keep it as the responsibility of the caller to

1. Provide the final path
2. do any "slugify" of the path

That being said, we should leave the code in place for `command_to_filename` to make it easier on the caller.

michalis1 replied:
I’ve updated the code so that the caller provides the full file path instead of the Git repository object and the relative path, for both get_command and get_commands, to ensure consistency.

itdependsnetworks commented on docs/task/task.md:
```diff
@@ -104,6 +104,44 @@ class DispatcherMixin:
+2. `obj.get_config_context()["offline_commands"]` — if it exists and is a valid boolean value.
+3. `cls.offline_commands` — the default class attribute defined in `NetmikoDefault`, which defaults to `False`.
+
+When enabled, the dispatcher attempts to read the expected command output from the filesystem instead of executing the command on a live device. This requires the output files to be named in a filesystem-safe format.
```
```suggestion
When enabled, the dispatcher attempts to read the expected command output from the filesystem (via the keyword arguments git_repo_obj, commands_folder_name, and command_relative_path) instead of executing the command on a live device. This requires the output files to be named in a filesystem-safe format.
```

jvanderaa commented on nornir_nautobot/plugins/tasks/dispatcher/mikrotik_routeros.py:
Why are we changing only `mikrotik_routeros.py` on this?

michalis1 replied:
It’s the only one extending `NetmikoDefault`, which is the chosen dispatcher for retrieving ‘offline’ command outputs from a Git repository.

itdependsnetworks commented on nornir_nautobot/plugins/tasks/dispatcher/default.py:
```diff
@@ -605,62 +624,176 @@ def merge_config(  # pylint: disable=too-many-positional-arguments
+        return Result(host=task.host, result=command_raw)
+
+    @classmethod
+    def get_command(
```
ping.

michalis1 replied:
I've already updated this one.

michalis1 commented on nornir_nautobot/plugins/tasks/dispatcher/default.py:
```diff
@@ -605,62 +624,176 @@ def merge_config(  # pylint: disable=too-many-positional-arguments
+            commands_folder_name: Name of the folder within the Git repository where command outputs are stored.
+            command_relative_path: The path to the command output file located under the command_outputs folder in the Git repository.
+        """
+        logger.debug(f"Executing get_command for {task.host.name} on {task.host.platform}")
```
done

michalis1 commented on nornir_nautobot/plugins/tasks/dispatcher/mikrotik_routeros.py:
```diff
@@ -102,7 +102,7 @@ class NetmikoMikrotikRouteros(NetmikoDefault):
 
     @classmethod
-    def get_config(  # pylint: disable=R0913,R0914,too-many-positional-arguments
+    def get_config(  # pylint: disable=R0913,R0914,too-many-positional-arguments,arguments-differ
```
done

michalis1 commented on nornir_nautobot/plugins/tasks/dispatcher/default.py:
```diff
@@ -605,62 +624,176 @@ def merge_config(  # pylint: disable=too-many-positional-arguments
-                logger.error(error_msg, extra={"object": obj})
-                raise NornirNautobotException(error_msg)
+            if cls._offline_commands(obj):
+                logger.info("Getting Offline commands from {kwargs.get('git_repo_obj')}", extra={"object": obj})
```
done
